clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java

```diff
@@ -609,7 +607,22 @@ private void handleProduceResponse(ClientResponse response, Map<TopicPartition,
                                 .collect(Collectors.toList()),
                             p.errorMessage(),
                             p.currentLeader());
-                    ProducerBatch batch = batches.get(tp);
+                    ProducerBatch batch = null;
+                    // Version 13 drop topic name and add support to topic id.
+                    // We need to find batch based on topic id and partition index only as
+                    // topic name in the response might be empty.
+                    List<ProducerBatch> matchedBatchesForTopicId = batches.entrySet().stream()
+                            .filter(entry -> entry.getKey().same(new TopicIdPartition(r.topicId(), p.index(), r.name())))
+                            .map(Map.Entry::getValue)
+                            .collect(Collectors.toList());
+
+                    if (matchedBatchesForTopicId.size() > 1) {
+                        matchedBatchesForTopicId.forEach(matchedBatch ->
+                                failBatch(matchedBatch, new RuntimeException("More than one batch with same topic id and partition."), false));
```

Comment:
If this is unexpected, we want to throw IllegalStateException as in other places in this file

Reply:
Removed this now in favour of using first find. I am just thinking should we in general return IllegalStateException if batch is null?